compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+
+   doneLabel->setEndInternalControlFlow();
+   generateDepLabelInstruction(cg, TR::InstOpCode::label, node, doneLabel, conditions);
+
+   cg->decReferenceCount(valueAddrChild);
+   cg->stopUsingRegister(valueAddrReg);
+   cg->stopUsingRegister(cndReg);
+
+   if (tempReg)
+      {
+      cg->stopUsingRegister(tempReg);
+      }
+
+   if (deltaChild)
+      {
+      cg->decReferenceCount(deltaChild);
+      }
+
+   if (isGetThenUpdate)
+      {
+      //for Get And Op, we will store the result in the result register
+      cg->stopUsingRegister(deltaReg);
+      node->setRegister(resultReg);
+      return resultReg;
+      }
+   else
+      {
+      //for Op And Get, we will store the return value in the delta register
+      //we no longer need the result register
+      cg->stopUsingRegister(resultReg);
+      node->setRegister(deltaReg);
+      return deltaReg;
+      }
+   }
+
+bool OMR::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&resultReg)
+   {
+   TR::CodeGenerator *cg = self();
+   TR::Compilation *comp = cg->comp();
+   TR::SymbolReference* symRef = node->getSymbolReference();
+   bool doInline = false;
+
+   if (symRef && symRef->getSymbol()->castToMethodSymbol()->isInlinedByCG())
+      {
+      bool isAddOp = false;
+      bool isLong = false;
+      bool isGetThenUpdate = false;
+
+      if (comp->getSymRefTab()->isNonHelper(symRef, TR::SymbolReferenceTable::atomicAdd32BitSymbol))
+         {
+         isAddOp = true;
+         isLong = false;
+         isGetThenUpdate = false;
+         doInline = true;
+         }
+      else if (comp->getSymRefTab()->isNonHelper(symRef, TR::SymbolReferenceTable::atomicAdd64BitSymbol))
+         {
+         isAddOp = true;
+         isLong = true;
+         isGetThenUpdate = false;
+         doInline = true;
+         }
+      else if (comp->getSymRefTab()->isNonHelper(symRef, TR::SymbolReferenceTable::atomicFetchAndAdd32BitSymbol))
+         {
+         isAddOp = true;
+         isLong = false;
+         isGetThenUpdate = true;
+         doInline = true;
+         }
+      else if (comp->getSymRefTab()->isNonHelper(symRef, TR::SymbolReferenceTable::atomicFetchAndAdd64BitSymbol))
+         {
+         isAddOp = true;
+         isLong = true;
+         isGetThenUpdate = true;
+         doInline = true;
+         }
+      else if (comp->getSymRefTab()->isNonHelper(symRef, TR::SymbolReferenceTable::atomicSwap32BitSymbol))
+         {
+         isAddOp = false;
+         isLong = false;
+         isGetThenUpdate = true;
+         doInline = true;
+         }
+      else if (comp->getSymRefTab()->isNonHelper(symRef, TR::SymbolReferenceTable::atomicSwap64BitSymbol))
+         {
+         isAddOp = false;
+         isLong = true;
+         isGetThenUpdate = true;
+         doInline = true;
+         }
+
+      if (doInline)
+         {
+         resultReg = inlineSimpleAtomicUpdate(node, isAddOp, isLong, isGetThenUpdate, cg);
+         }
+      }
+
+   return doInline;
+   }
+
```

Comment:
High-level issue:  this will not work for 32bit JVM.  i.e.  this is assuming a 64bit JVM.

Reply:
@zl-wang Julian, thanks for catching this.  Just so I'm clear, can this code be inlined on a 32 bit JVM when it's working with 32 bit operands, and inlined on a 64 bit JVM when it's working with either 32 bit and 64 bit operands?  Or can it only be inlined on a 64 JVM ever?

Something I just noticed this morning, I was playing with `AtomicLong.getAndAdd` and `AtomicInteger.getAndAdd` without my changes for these new intrinsics, and it looks like inline code is never generated for `AtomicLong.getAndAdd` for 32 bit or 64 bit JVMs, and always generated for `AtomicInteger.getAndAdd` for both 32 bit and 64 bit JVMs.  Is that expected?

Reply:
@hzongaro I didn't expect that.  I expected existing JVM can inline CAS of both 32/64 operands in both 32/64 bit JVM.  The only assumption is the underlying CPU is 64bit (this has been true since java7.1 when we discontinued support of 32bit hardware).  The case for inlining CAS of 64bit operand in 32bit JVM needs special care though ... you can assemble 64bit operand in a register within the CAS range only, depending on the fact that CAS will definitely fail if there was a signal delivered in the between (i.e.  meaning the upper 32bit of the register can be trashed anytime).

Reply:
@zl-wang Julian, I've been working on a revision that handles 64-bit values running in a 32-bit JVM, but it makes the code quite ugly.  I’ll pull that out as a separate pull request, and I’ll let you decide whether it’s an important enough case to handle, as well as point out any problems with the way that I tackled the problem.